### PR TITLE
WIP: Integration of the zephyr-xenlib to application

### PR DESCRIPTION
**Pull request by oleksiimoisieiev**
Also there were integrated xen binary and bundle creation for the testing purposes.
Building and running:
west build -b xenvm-qemu -p always
west build -t fix_dtb
west build -t run

Test xrun interface:
xrun run -c t -b /lfs -s 0

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,11 +27,18 @@ west zephyr-export
 cd aos_core_zephyr
 
 west build -b ${BOARD} -p auto
+west build -t fix_dtb
```
Comment (al1img) [inline README.md:30]: It looks like some kind of WA, doesn't it? Please add explanation what this target does and when we need to use it.

File: scripts/fix_dtb.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+TEMPLATE=$1
+TEMPLATE=${TEMPLATE:-gicv3.tmpl}
+OUTPUT=$2
+OUTPUT=${OUTPUT:-$IMAGE_PATH/virt_gicv3.dtb}
+IMAGE=$3
+IMAGE=${IMAGE:-$IMAGE_PATH/zephyr.bin}
+
+SIZE=$(printf "0x%x\n" $(stat -c %s $IMAGE))
+sed  s/\$SIZE/$SIZE/g < $TEMPLATE | dtc -o $OUTPUT
+#dtc gicv2.dts -o $OUTPUT
```
Comment (al1img) [inline scripts/fix_dtb.sh:14]: Do you need this comment?

File: prj.conf [hunk 3]
```diff
@@ -27,3 +27,6 @@ CONFIG_MP_NUM_CPUS=1
 CONFIG_XEN_STORE_SRV=y
 CONFIG_XEN_DOMAIN_MANAGEMENT=y
 CONFIG_XEN_SHELL=y
+CONFIG_XEN_LIBFDT=y
+CONFIG_XRUN=y
+CONFIG_JSON_LIBRARY=y
```
Comment (al1img) [inline prj.conf:32]: XRUN, JSON lib should be in separate block with appropriate comment.

File: src/domain_bins.S [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+__img_unikraft_start:
+.incbin "helloworld_xen-arm64"
+__img_unikraft_end:
+
+.section .data.dtb_unikraft,"a"
+
+GDATA(__dtb_unikraft_start)
+GDATA(__dtb_unikraft_end)
+__dtb_unikraft_start:
```
Comment (al1img) [inline src/domain_bins.S:22]: Could you add empty line after GDATA as in above case?

File: CMakeLists.txt [hunk 5]
```diff
@@ -97,6 +99,23 @@ ExternalProject_Add(
     BUILD_BYPRODUCTS ${AOSCORE_BUILD_DIR}/lib/libaossmcpp.a ${AOSCORE_BUILD_DIR}/lib/libaosiamcpp.a
 )
 
+set(XEN_SRCDIR ${CMAKE_CURRENT_SOURCE_DIR}/../xen)
```
Comment (al1img) [inline CMakeLists.txt:102]: small letter for local vars

File: CMakeLists.txt [hunk 6]
```diff
@@ -97,6 +99,23 @@ ExternalProject_Add(
...
+set(XEN_BINDIR ${CMAKE_CURRENT_BINARY_DIR}/hyp)
+
+ExternalProject_Add(
```
Comment (al1img) [inline CMakeLists.txt:105]: This is relevant only to specific board or arch. We should not build it for all boards.

File: src/main.cpp [hunk 7]
```diff
@@ -4,6 +4,7 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
+#define DEBUG
```
Comment (al1img) [inline src/main.cpp:7]: Why? DEBUG should be enabled (if required) in project config not in the code.

File: src/main.cpp [hunk 8]
```diff
@@ -14,11 +15,66 @@
 #include "logger/logger.hpp"
 #include "version.hpp"
 
+#if defined(DEBUG)
```
Comment (al1img) [inline src/main.cpp:18]: It should not be in app main. Test case should be created for such a purpose.

File: west.yml [hunk 9]
```diff
@@ -21,6 +21,8 @@ manifest:
     - name: zephyr
       url-base: https://github.com/zephyrproject-rtos
 
+    - name: amoi
```
Comment (al1img) [inline west.yml:24]: Is this only for debug purpose?

File: boards/xenvm-qemu.overlay [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+			read-size = <16>;
+			prog-size = <16>;
+			cache-size = <64>;
+			lookahead-size = <32>;
+			block-cycles = <512>;
+		};
+	};
+	flashcontroller0: flashcontroller {
+		compatible = "zephyr,sim-flash";
+		label = "FLASH_SIMULATOR";
```
Comment (xakep-amatop) [inline boards/xenvm-qemu.overlay:21]: some notes from Zephyr documentation:
```
Earlier versions of Zephyr made frequent use of the label property, which is distinct from the 
tandard node label. Use of the label property in new devicetree bindings, as well as use of the
DT_LABEL macro in new code, are actively discouraged. Label properties continue to persist
for historical reasons in some existing bindings and overlays, but should not be used in new
bindings or device implementations.
```

moreover, I have seen warnings form `west` that `label` property is a deprecated during compilation of a project where I had DTS labels.

**Comment by xakep-amatop** [comment]
@oleksiimoisieiev Should you add a license preamble to all new files?

File: src/main.cpp [hunk 11]
```diff
@@ -14,11 +15,66 @@
...
+extern void init_root();
+
+static void prepare_configs()
+{
+	int ret;
+
+	init_root();
+
+	printk("dtb = %p\n", (void *)__dtb_unikraft_start);
+	ret = write_file("/lfs", "config.json", json_msg, sizeof(json_msg));
```
Comment (xakep-amatop) [inline src/main.cpp:56]: (question) do you need to continue execution in case when one of `write_file` calls fail?

File: src/main.cpp [hunk 12]
```diff
@@ -14,11 +15,66 @@
...
+
+} /* extern "C" */
+#endif /* DEBUG */
+
 int main(void)
 {
     printk("*** Aos zephyr application: %s ***\n", AOS_ZEPHYR_APP_VERSION);
     printk("*** Aos core library: %s ***\n", AOS_CORE_VERSION);
 
+	prepare_configs();
```
Comment (xakep-amatop) [inline src/main.cpp:76]: (question) Why do you need so strange alignment?

File: scripts/gicv2.tmpl [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+			reg = <0x00>;
+			enable-method = "psci";
+			compatible = "arm,cortex-a57";
+			device_type = "cpu";
+		};
+	};
+
+	timer {
+		interrupts = <0x01 0x0d 0xf04 0x01 0x0e 0xf04 0x01 0x0b 0xf04 0x01 0x0a 0xf04>;
+		always-on;
```
Comment (xakep-amatop) [inline scripts/gicv2.tmpl:375]: (question) does timer have `always-on` property in Zephyr?

**Comment by xakep-amatop** [comment]
can we avoid somehow a lot of copy-pasted code inside giv2.tmpl & fiv3.tmpl? as for me, diff isn't so big:
```
index 7d15a98..16c2c56 100644
--- a/1.dts
+++ b/2.dts
@@ -301,7 +301,6 @@
                #interrupt-cells = <0x01>;
                ranges = <0x1000000 0x00 0x00 0x00 0x3eff0000 0x00 0x10000 0x2000000 0x00 0x10000000 0x00 0x10000000 0x00 0x2eff0000 0x3000000 0x80 0x00 0x80 0x00 0x80 0x00>;
                reg = <0x40 0x10000000 0x00 0x10000000>;
-               msi-parent = <&intc_v2m>;
                dma-coherent;
                bus-range = <0x00 0xff>;
                linux,pci-domain = <0x00>;
@@ -328,28 +327,21 @@
        };
 
        pmu {
-               interrupts = <0x01 0x07 0xf04>;
+               interrupts = <0x01 0x07 0x04>;
                compatible = "arm,armv8-pmuv3";
        };
 
        intc: intc@8000000 {
                interrupts = <0x01 0x09 0x04>;
-               reg = <0x00 0x8000000 0x00 0x10000
-                      0x00 0x8010000 0x00 0x10000
-                      0x00 0x8030000 0x00 0x10000
-                      0x00 0x8040000 0x00 0x10000>;
-               compatible = "arm,cortex-a15-gic";
+               reg = <0x00 0x8000000 0x00 0x20000
+                      0x00 0x80a0000 0x00 0xf60000>;
+               #redistributor-regions = <0x01>;
+               compatible = "arm,gic-v3";
                ranges;
                #size-cells = <0x02>;
                #address-cells = <0x02>;
                interrupt-controller;
                #interrupt-cells = <0x03>;
-
-               intc_v2m: v2m@8020000 {
-                       reg = <0x00 0x8020000 0x00 0x1000>;
-                       msi-controller;
-                       compatible = "arm,gic-v2m-frame";
-               };
        };
 
        flash@0 {
@@ -368,10 +360,11 @@
                        compatible = "arm,cortex-a57";
                        device_type = "cpu";
                };
+
        };
 
        timer {
-               interrupts = <0x01 0x0d 0xf04 0x01 0x0e 0xf04 0x01 0x0b 0xf04 0x01 0x0a 0xf04>;
+               interrupts = <0x01 0x0d 0x04 0x01 0x0e 0x04 0x01 0x0b 0x04 0x01 0x0a 0x04>;
                always-on;
                compatible = "arm,armv8-timer", "arm,armv7-timer";
        };
@@ -385,6 +378,7 @@
 
        chosen {
                stdout-path = "/pl011@9000000";
+               kaslr-seed = <0x6c530d2 0x3d747be8>;
                #address-cells = <1>;
                #size-cells = <1>;
                module@0 {
(END)
        pmu {
-               interrupts = <0x01 0x07 0xf04>;
+               interrupts = <0x01 0x07 0x04>;
                compatible = "arm,armv8-pmuv3";
        };
 
        intc: intc@8000000 {
                interrupts = <0x01 0x09 0x04>;
-               reg = <0x00 0x8000000 0x00 0x10000
-                      0x00 0x8010000 0x00 0x10000
-                      0x00 0x8030000 0x00 0x10000
-                      0x00 0x8040000 0x00 0x10000>;
-               compatible = "arm,cortex-a15-gic";
+               reg = <0x00 0x8000000 0x00 0x20000
+                      0x00 0x80a0000 0x00 0xf60000>;
+               #redistributor-regions = <0x01>;
+               compatible = "arm,gic-v3";
                ranges;
                #size-cells = <0x02>;
                #address-cells = <0x02>;
                interrupt-controller;
                #interrupt-cells = <0x03>;
-
-               intc_v2m: v2m@8020000 {
-                       reg = <0x00 0x8020000 0x00 0x1000>;
-                       msi-controller;
-                       compatible = "arm,gic-v2m-frame";
-               };
        };
 
        flash@0 {
@@ -368,10 +360,11 @@
                        compatible = "arm,cortex-a57";
                        device_type = "cpu";
                };
+
        };
 
        timer {
-               interrupts = <0x01 0x0d 0xf04 0x01 0x0e 0xf04 0x01 0x0b 0xf04 0x01 0x0a 0xf04>;
+               interrupts = <0x01 0x0d 0x04 0x01 0x0e 0x04 0x01 0x0b 0x04 0x01 0x0a 0x04>;
                always-on;
                compatible = "arm,armv8-timer", "arm,armv7-timer";
        };
@@ -385,6 +378,7 @@
 
        chosen {
                stdout-path = "/pl011@9000000";
+               kaslr-seed = <0x6c530d2 0x3d747be8>;
                #address-cells = <1>;
                #size-cells = <1>;
                module@0 {
```

File: scripts/gicv2.tmpl (hunk 13 above)
Comment (oleksiimoisieiev) [reply scripts/gicv2.tmpl:375]: This changes are the reference for integration to aos-app. Thank you for review, but gic_v2/v3 should be added as different MR.

**Comment by oleksiimoisieiev** [comment]
@xakep-amatop Sorry I didn't mentioned it in MR description. i've Marked MR as WIP and we've used it mostly as reference to continue implementation and we have no plans to merge it.

**Comment by oleksiimoisieiev** [comment]
Closing this MR as not relevant anymore.

**Pull request closed without merging**